File: scripts/day1/day1_scripts_R/Day1_Application1_ClinicalTrials.R
# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 1 - From web source to raw dataset
# Application 1: Health Sciences - ClinicalTrials.gov API
# =============================================================================
# Goal: Extract raw structured data from an online source (API) and cache it locally.

library(httr2)
library(jsonlite)
library(fs)

fetch_clinical_trials_data <- function() {
  cache_dir <- path(path_dir(path_dir(getwd())), "data", "cache")
  dir_create(cache_dir)
  cache_file <- path(cache_dir, "day1_app1_clinicaltrials_raw.json")
  
  if (file_exists(cache_file)) {
    cat("Loading cached data from", cache_file, "\n")
    raw_data <- fromJSON(cache_file, simplifyVector = FALSE)
  } else {
    cat("Fetching fresh data from ClinicalTrials.gov API...\n")
    url <- "https://clinicaltrials.gov/api/v2/studies"
    
    req <- request(url) %>%
      req_url_query(
        `query.cond` = "diabetes",
        pageSize = 50,
        fields = "NCTId,Condition,BriefTitle,OverallStatus,LeadSponsorName"
      )
    
    resp <- req_perform(req)
    raw_data <- resp_body_json(resp, simplifyVector = FALSE)
    
    write_json(raw_data, cache_file, pretty = TRUE, auto_unbox = TRUE)
    cat("Saved raw data to", cache_file, "\n")
  }
  
  if (!is.null(raw_data$studies) && length(raw_data$studies) > 0) {
    cat("\nExample of raw extracted record:\n")
    print(toJSON(raw_data$studies[[1]], pretty = TRUE, auto_unbox = TRUE))
  } else {
    cat("No studies found.\n")
  }
}

fetch_clinical_trials_data()

File: scripts/day1/day1_scripts_Python/Day1_Application1_ClinicalTrials.py
import os
import json
import pandas as pd
import requests

# =============================================================================
# Instats Seminar: Web Scraping and Data Collection
# Day 1 - From web source to raw dataset
# Application 1: Health Sciences - ClinicalTrials.gov API
# =============================================================================
# Goal: Extract raw structured data from an online source (API) and cache it locally.
# This script demonstrates how to identify a source, request data, and save the raw output.

def fetch_clinical_trials_data():
    """
    Fetches clinical trial data related to 'diabetes' from the ClinicalTrials.gov API.
    Saves the raw JSON response to a local cache file.
    """
    cache_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data', 'cache')
    os.makedirs(cache_dir, exist_ok=True)
    cache_file = os.path.join(cache_dir, 'day1_app1_clinicaltrials_raw.json')

    # Check if we already have the data cached
    if os.path.exists(cache_file):
        print(f"Loading cached data from {cache_file}")
        with open(cache_file, 'r', encoding='utf-8') as f:
            raw_data = json.load(f)
    else:
        print("Fetching fresh data from ClinicalTrials.gov API...")
        # ClinicalTrials.gov v2 API endpoint for study search
        # We search for 'diabetes' and limit to 50 results for manageable size
        url = "https://clinicaltrials.gov/api/v2/studies"
        params = {
            "query.cond": "diabetes",
            "pageSize": 50,
            "fields": "NCTId,Condition,BriefTitle,OverallStatus,LeadSponsorName"
        }
        
        response = requests.get(url, params=params)
        response.raise_for_status()  # Check for request errors
        raw_data = response.json()
        
        # Save to cache
        with open(cache_file, 'w', encoding='utf-8') as f:
            json.dump(raw_data, f, indent=4)
        print(f"Saved raw data to {cache_file}")

    # Display the first study as an example of raw output
    if 'studies' in raw_data and len(raw_data['studies']) > 0:
        first_study = raw_data['studies'][0]
        print("\nExample of raw extracted record:")
        print(json.dumps(first_study, indent=2))
    else:
        print("No studies found.")

if __name__ == "__main__":
    fetch_clinical_trials_data()
